File: backend/featurePacks/ftpPack.ts
import * as ftp from 'basic-ftp';
import fs from 'fs';
import path from 'path';

export interface FtpDownloadProgress {
  bytesDownloaded: number;
  totalBytes: number;
  percent: number;
  speed: string;
  eta: string;
}

export class FtpPack {
  private client: ftp.Client;
  private aborted = false;

  constructor() {
    this.client = new ftp.Client();
    this.client.ftp.verbose = false;
  }

  /**
   * Parse an FTP/FTPS URL
   */
  public parseUrl(urlStr: string): {
    host: string;
    port: number;
    user?: string;
    password?: string;
    secure: boolean;
    remotePath: string;
    fileName: string;
  } {
    const parsed = new URL(urlStr);
    const isSecure = parsed.protocol === 'ftps:';
    const remotePath = decodeURIComponent(parsed.pathname);
    const fileName = path.basename(remotePath) || 'ftp_download';

    return {
      host: parsed.hostname,
      port: parsed.port ? parseInt(parsed.port, 10) : (isSecure ? 990 : 21),
      user: parsed.username || 'anonymous',
      password: parsed.password || 'anonymous@vortex.org',
      secure: isSecure,
      remotePath,
      fileName
    };
  }

  /**
   * Download a file from FTP server with progress
   */
  public async downloadFile(
    urlStr: string,
    destinationDir: string,
    onProgress?: (prog: FtpDownloadProgress) => void
  ): Promise<{ filePath: string; fileName: string; size: number }> {
    const info = this.parseUrl(urlStr);
    const targetFile = path.join(destinationDir, info.fileName);

    let totalBytes = 0;
    let bytesDownloaded = 0;
    let startTime = Date.now();
    let lastTime = Date.now();
    let lastBytes = 0;

    try {
      await this.client.access({
        host: info.host,
        port: info.port,
        user: info.user,
        password: info.password,
        secure: info.secure
      });

      // Probe remote size
      try {
        totalBytes = await this.client.size(info.remotePath);
      } catch (_) {
        totalBytes = 0;
      }

      this.client.trackProgress((info) => {
        if (this.aborted) {
          this.client.close();
          return;
        }

        bytesDownloaded = info.bytes;
        const now = Date.now();
        const elapsedSec = (now - startTime) / 1000;
        const deltaSec = (now - lastTime) / 1000;

        let speedStr = '-- B/s';
        if (deltaSec >= 0.5) {
          const bps = (bytesDownloaded - lastBytes) / deltaSec;
          speedStr = this.formatSpeed(bps);
          lastTime = now;
          lastBytes = bytesDownloaded;
        }

        let etaStr = '--';
        if (totalBytes > 0 && bytesDownloaded > 0 && elapsedSec > 0) {
          const overallBps = bytesDownloaded / elapsedSec;
          const remainingBytes = totalBytes - bytesDownloaded;
          const remSec = remainingBytes / overallBps;
          etaStr = this.formatDuration(remSec);
        }

        const percent = totalBytes > 0 ? Math.min(100, Math.round((bytesDownloaded / totalBytes) * 100)) : 0;

        if (onProgress) {
          onProgress({
            bytesDownloaded,
            totalBytes,
            percent,
            speed: speedStr,
            eta: etaStr
          });
        }
      });

      // Stream file to disk
      await this.client.downloadTo(targetFile, info.remotePath);
      this.client.trackProgress(); // Reset tracking

      const stats = fs.existsSync(targetFile) ? fs.statSync(targetFile) : null;
      return {
        filePath: targetFile,
        fileName: info.fileName,
        size: stats?.size || totalBytes || bytesDownloaded
      };
    } finally {
      this.client.close();
    }
  }

  public abort(): void {
    this.aborted = true;
    this.client.close();
  }

  private formatSpeed(bps: number): string {
    if (bps >= 1024 * 1024) return `${(bps / (1024 * 1024)).toFixed(2)} MB/s`;
    if (bps >= 1024) return `${(bps / 1024).toFixed(1)} KB/s`;
    return `${Math.round(bps)} B/s`;
  }

  private formatDuration(sec: number): string {
    if (!sec || !isFinite(sec) || sec < 0) return '--';
    const s = Math.round(sec);
    const m = Math.floor(s / 60);
    const rSec = s % 60;
    if (m >= 60) {
      const h = Math.floor(m / 60);
      const rMin = m % 60;
      return `${h}h ${rMin}m`;
    }
    return `${m}m ${rSec}s`;
  }
}

File: backend/feature_packs/ftp_pack.py
import os
import time
from urllib.parse import urlparse, unquote
from ftplib import FTP, FTP_TLS
from typing import Callable, Optional, Dict, Any

class FtpPack:
    def __init__(self):
        self.client: Optional[FTP] = None
        self.is_cancelled = False

    def parse_url(self, raw_url: str) -> Dict[str, Any]:
        parsed = urlparse(raw_url)
        is_secure = parsed.scheme.lower() == 'ftps'
        host = parsed.hostname or ''
        port = parsed.port or (990 if is_secure else 21)
        user = unquote(parsed.username) if parsed.username else 'anonymous'
        password = unquote(parsed.password) if parsed.password else 'anonymous@vortex.app'
        file_path = unquote(parsed.path).lstrip('/')
        file_name = os.path.basename(file_path) or 'ftp_download.bin'

        return {
            'isSecure': is_secure,
            'host': host,
            'port': port,
            'user': user,
            'password': password,
            'path': file_path,
            'fileName': file_name
        }

    def cancel(self):
        self.is_cancelled = True
        if self.client:
            try:
                self.client.abort()
                self.client.close()
            except Exception:
                pass

    def download_file(
        self,
        url: str,
        destination_dir: str,
        on_progress: Optional[Callable[[Dict[str, Any]], None]] = None
    ) -> Dict[str, Any]:
        info = self.parse_url(url)
        os.makedirs(destination_dir, exist_ok=True)
        out_path = os.path.join(destination_dir, info['fileName'])

        # Connect
        ftp_cls = FTP_TLS if info['isSecure'] else FTP
        ftp = ftp_cls(timeout=15)
        self.client = ftp

        try:
            ftp.connect(info['host'], info['port'])
            ftp.login(info['user'], info['password'])
            if info['isSecure'] and isinstance(ftp, FTP_TLS):
                ftp.prot_p()

            # Get file size
            total_size = 0
            try:
                total_size = ftp.size(info['path']) or 0
            except Exception:
                total_size = 0

            downloaded = 0
            start_time = time.time()
            last_time = start_time
            last_bytes = 0

            with open(out_path, 'wb') as f:
                def handle_chunk(data: bytes):
                    nonlocal downloaded, last_time, last_bytes
                    if self.is_cancelled:
                        raise Exception("FTP Download cancelled by user.")
                    f.write(data)
                    downloaded += len(data)

                    now = time.time()
                    if on_progress and (now - last_time >= 0.5 or (total_size and downloaded == total_size)):
                        elapsed = now - last_time
                        speed_bps = (downloaded - last_bytes) / elapsed if elapsed > 0 else 0
                        last_time = now
                        last_bytes = downloaded

                        percent = int((downloaded / total_size) * 100) if total_size > 0 else 0
                        speed_mb = speed_bps / (1024 * 1024)
                        rem_bytes = max(0, total_size - downloaded)
                        eta_s = int(rem_bytes / speed_bps) if speed_bps > 0 else 0

                        on_progress({
                            'bytesDownloaded': downloaded,
                            'totalBytes': total_size,
                            'percent': percent,
                            'speed': f"{speed_mb:.1f} MB/s",
                            'eta': f"{eta_s}s" if eta_s > 0 else "--"
                        })

                ftp.retrbinary(f"RETR {info['path']}", handle_chunk, blocksize=64 * 1024)

            return {'filePath': out_path, 'bytesDownloaded': downloaded}

        finally:
            try:
                ftp.quit()
            except Exception:
                try:
                    ftp.close()
                except Exception:
                    pass
            self.client = None
